modules/frontend/pipeline/sync_handler_cache.go

```diff
@@ -0,0 +1,165 @@
+	return MiddlewareFunc(func(next http.RoundTripper) http.RoundTripper {
+		return cachingWare{
+			next:  next,
+			cache: newFrontendCache(cacheProvider, role, logger),
+		}
+	})
+}
+
+type cachingWare struct {
+	next  http.RoundTripper
+	cache *frontendCache
+}
+
+// RoundTrip implements http.RoundTripper
+func (c cachingWare) RoundTrip(req *http.Request) (*http.Response, error) {
+	// short circuit everything if there cache is no cache
+	if c.cache == nil {
+		return c.next.RoundTrip(req)
+	}
+
+	// extract cache key
+	key, ok := req.Context().Value(cacheKey).(string)
+	if ok && len(key) > 0 {
+		body := c.cache.fetchBytes(key)
+		if len(body) > 0 {
+			return &http.Response{
+				StatusCode: http.StatusOK,
+				Status:     http.StatusText(http.StatusOK),
+				Body:       io.NopCloser(bytes.NewBuffer(body)),
+			}, nil
+		}
+	}
+
+	resp, err := c.next.RoundTrip(req)
+	// do not cache if there was an error
+	if err != nil {
+		return resp, err
+	}
+
+	// do not cache if response is not HTTP 2xx
+	if !shouldCache(resp.StatusCode) {
+		return resp, nil
+	}
+
+	if len(key) > 0 {
+		// cache the response
+		//  todo: currently this is blindly caching any 200 status codes. it would be a bug, but it's possible for a querier
+		//  to return a 200 status code with a response that does not parse as the expected type in the combiner.
+		//  technically this should never happen...
+		//  long term we should migrate the sync part of the pipeline to use generics so we can do the parsing early in the pipeline
+		//  and be confident it's cacheable.
+		b, err := io.ReadAll(resp.Body)
+
+		// reset the body so the caller can read it
+		resp.Body = io.NopCloser(bytes.NewBuffer(b))
+		if err != nil {
+			return resp, nil
+		}
+
+		c.cache.store(req.Context(), key, b)
+	}
+
+	return resp, nil
+}
+
+func shouldCache(statusCode int) bool {
+	return statusCode/100 == 2
+}
+
+func AddCacheKey(key string, req *http.Request) *http.Request {
+	return req.WithContext(context.WithValue(req.Context(), cacheKey, key))
+}
+
+type frontendCache struct {
+	c cache.Cache
+}
+
+func newFrontendCache(cacheProvider cache.Provider, role cache.Role, logger log.Logger) *frontendCache {
+	var c cache.Cache
+	if cacheProvider != nil {
+		c = cacheProvider.CacheFor(role)
+	}
+
+	level.Info(logger).Log("msg", "init frontend cache", "role", role, "enabled", c != nil)
+
+	if c == nil {
+		return nil
+	}
+
+	return &frontendCache{
+		c: c,
+	}
+}
+
+// store stores the response body in the cache. the caller assumes the responsibility of closing the response body
+func (c *frontendCache) store(ctx context.Context, key string, buffer []byte) {
+	if c.c == nil {
+		return
+	}
+
+	if key == "" {
+		return
+	}
+
+	if len(buffer) == 0 {
+		return
+	}
+
+	c.c.Store(ctx, []string{key}, [][]byte{buffer})
+}
+
+// fetch fetches the response body from the cache. the caller assumes the responsibility of closing the response body.
+func (c *frontendCache) fetch(key string, pb proto.Message) bool {
```

Comment:
This method looks unused except for the test, and wasn't sure about json+proto.Message.  Since we know it's a `proto.Message` we could store the binary representation in cache instead.

Reply:
yup. i do think there's an opportunity to make the sync part of the pipeline type aware with generics which would allow this, but we're not quite there yet